atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/mcas/MarkAndCasConsensusForgettingStore.java

```diff
@@ -82,23 +84,37 @@ public void mark(Set<Cell> cells) {
     }
 
     /**
-     * Atomically updates cells that have been marked. Throws {@code CheckAndSetException} if cell to update has not
-     * been marked. The MCAS calls to KVS are batched.
+     * Atomically updates cells that have been marked. The MCAS calls to KVS are batched.
+     *
+     * @return {@link AtomicUpdateResult} with success if the atomic update was successful. Else,
+     * {@link AtomicUpdateResult} with {@link KeyAlreadyExistsException} with detail if there is a value
+     * present against this key.
      */
     @Override
-    public void atomicUpdate(Cell cell, byte[] value) throws KeyAlreadyExistsException {
-        autobatcher.apply(ImmutableCasRequest.of(cell, inProgressMarkerBuffer, ByteBuffer.wrap(value)));
+    public AtomicUpdateResult atomicUpdate(Cell cell, byte[] value) {
+        ImmutableCasRequest casRequest = ImmutableCasRequest.of(cell, inProgressMarkerBuffer, ByteBuffer.wrap(value));
+        return getResultForFuture(casRequest);
     }
 
     /**
      * This call serially delegated to {@link MarkAndCasConsensusForgettingStore#atomicUpdate(Cell, byte[])}
      * and does not guarantee atomicity across cells.
      * The MCAS calls to KVS are batched and hence, in practice, it is possible that the group of cells is served
      * atomically.
-     * */
+     */
     @Override
-    public void atomicUpdate(Map<Cell, byte[]> values) throws KeyAlreadyExistsException {
-        values.forEach(this::atomicUpdate);
+    public AtomicUpdateResult atomicUpdate(Map<Cell, byte[]> values) {
+        ImmutableList.Builder<Cell> committedKeys = ImmutableList.builder();
+        ImmutableList.Builder<Cell> existingKeys = ImmutableList.builder();
+
+        values.entrySet().stream()
+                .map(entry -> atomicUpdate(entry.getKey(), entry.getValue()))
+                .forEach(updateResult -> {
+                    committedKeys.addAll(updateResult.knownSuccessfullyCommittedKeys());
+                    existingKeys.addAll(updateResult.existingKeys());
+                });
```

Comment:
No, this exception handling does not work. Each individual can have its future fail and not due to keyAlreadyExists exception.